File: test/src/pyspark/group_top_n.py
#!/usr/bin/python
# encoding:utf-8

# Copyright (c) 2017 Baidu, Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# usage: spark-submit wordCount.py

from pyspark.sql import SQLContext
from pyspark import SparkContext, SparkConf
import re,sys
from operator import add
from pyspark.serializers import MarshalSerializer

def main():
    input_path = sys.argv[1]
    output_path = sys.argv[2]
    conf = SparkConf().setAppName("PySpark: group_top_n ")#.set("spark.defalut.parallelism","250")
    sc = SparkContext(conf = conf)
    #sc = SparkContext(conf = conf,serializer=MarshalSerializer())
    sqlCtx = SQLContext(sc)

    lines = sc.textFile(input_path,250)
    res = lines.map(lambda line: re.split(' ', line))\
            .groupByKey()\
            .mapValues(lambda x :  sorted(x,reverse=True)[:5])\
            .flatMapValues(lambda x:x)\
            .map(lambda x: str(x[0]) +" "+ str(x[1]))

    res.saveAsTextFile(output_path)

if __name__ == '__main__':
    main()

File: test/src/scala/src/main/scala/com/baidu/inf/GroupTopN.scala
package com.baidu.inf

import org.apache.spark._
import org.apache.spark.SparkContext._

object GroupTopN {
  def main(args: Array[String]) {
    if (args.length < 3) {
      println("usage: spark-submit com.baidu.inf.WordCount <input> <output>")
      System.exit(1)
    }
    val conf = new SparkConf()
    conf.setAppName("group_top_n-" + System.getenv("USER"))

    val sc = new SparkContext(conf)

    // input file uri, could be any hdfs uri OR local fs uri.
    // for example:
    //  hdfs://baidu.com:1234/app/dc/spark/test.log
    //  file:///tmp/test.log
    // Note: local fs uri could only be used in spark local mode(set spark.master=local)
    val input = args(0)

    // output file uri, could be any hdfs uri OR local fs uri.
    val output = args(1)

    val topcount = args(2).toInt


    val file = sc.textFile(input,250)
    val counts = file.map(line => line.split(" "))
                     .map(x => (x(0),x(1)))
                     .groupByKey()
                     .mapValues(iter => iter.toList.sorted(Ordering[String].reverse).take(topcount))
                     .flatMapValues( iter =>iter )
                     .map(x => x._1 + " " + x._2 )
    //counts.foreach(println)
    counts.saveAsTextFile(output)

    System.exit(0)
  }
}
